File: bzz-spark/src/main/java/com/spark/study/ActionOperation.java
package com.spark.study;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.VoidFunction;

import java.util.Arrays;
import java.util.List;

/**
 * @Title: ActionOperation
 * @ProjectName bzz-cloud
 * @Auther: clouder
 * @Date: 2018-07-13-18:10
 * @Description:
 */
public class ActionOperation {
	/**
	 * @Description: TODO
	 * @param
	 * @return void
	 * @throws
	 * @author 624003618@qq.com
	 * @date 2018-07-13 18:23
	 */
	private static void reduce(){
		SparkConf conf = new SparkConf().setAppName("ActionOperation").setMaster("local");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
		
		JavaRDD<Integer> numbers = jsc.parallelize(numberList);
		Integer reduce = numbers.reduce(new Function2<Integer, Integer, Integer>() {
			public Integer call(Integer v1, Integer v2) throws Exception {
				return v1 + v2;
			}
		});
		System.out.println(reduce);
		
		
		
		jsc.close();
	}
	
	private static void map(){
		SparkConf conf = new SparkConf().setAppName("ActionOperation").setMaster("local");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
		
		JavaRDD<Integer> numbers = jsc.parallelize(numberList);
		JavaRDD<Integer> map = numbers.map(new Function<Integer, Integer>() {
			
			@Override
			public Integer call(Integer v) throws Exception {
				return v * 2;
			}
		});
		map.foreach(new VoidFunction<Integer>() {
			@Override
			public void call(Integer v) throws Exception {
				System.out.println(v);
			}
		});
		
		
		jsc.close();
	}
	
	private static void count(){
		SparkConf conf = new SparkConf().setAppName("ActionOperation").setMaster("local");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
		
		JavaRDD<Integer> numbers = jsc.parallelize(numberList);
		long count = numbers.count();
		System.out.println(count);
		
		jsc.close();
	}
	private static void take(){
		SparkConf conf = new SparkConf().setAppName("ActionOperation").setMaster("local");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
		
		JavaRDD<Integer> numbers = jsc.parallelize(numberList);
		List<Integer> lists = numbers.take(3);
		System.out.println(lists.toString());
		jsc.close();
	}
	private static void saveASText(){
		SparkConf conf = new SparkConf().setAppName("ActionOperation");
		conf.set("spark.testing.memory", "2147480000");
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		List<Integer> numberList = Arrays.asList(1,2,3,4,5,6,7,8,9,10);
		
		JavaRDD<Integer> numbers = jsc.parallelize(numberList);
		Integer reduce = numbers.reduce(new Function2<Integer, Integer, Integer>() {
			public Integer call(Integer v1, Integer v2) throws Exception {
				return v1 + v2;
			}
		});
		numbers.saveAsTextFile("hdfs://hadoop-cloud:8020/spark/numbers.txt");
		
		System.out.println(reduce);
		
		jsc.close();
	}
	public static void main(String[] args) {
		/*reduce();*/
		/*map();*/
		/*count();*/
		saveASText();
	}
	
}

File: bzz-spark/src/main/scala/com/spark/scala/study/ActionOperation.scala
package com.spark.scala

import org.apache.spark.{SparkConf, SparkContext}

/**
  * @Title: package
  * @ProjectName bzz-cloud
  * @Auther: clouder
  * @Date: 2018-07-13-18:50
  * @Description:
  */
object ActionOperation {

  /**
   * @param $ {tags}
   * @return ${return_type}
   * @throws
   * @author ${USER}
   * @date 2018-07-13 18:59
   */

  def reduce(): Unit ={
    var sparkConf = new SparkConf().setAppName("ActionOperation").setMaster("local")

    var sparkContext = new SparkContext(sparkConf)

    val nums : List[Int] = List(1, 2, 3, 4,5,6,7,8,9,10)
    val numbers = sparkContext.parallelize(nums)
    val sum = numbers.reduce((v1, v2) => {
      v1 + v2
    })
    println(sum)
  }

  def map(): Unit ={
    val sparkConf = new SparkConf().setMaster("local").setAppName("ActionOperation")
    val sparkContext = new SparkContext(sparkConf)

    val nums : List[Int] = List(1, 2, 3, 4,5,6,7,8,9,10)

    val number = sparkContext.parallelize(nums)
    val map = number.map( v => v*2 )
    map.foreach(v => println(v))
  }
  def count(): Unit ={
    val sparkConf = new SparkConf().setMaster("local").setAppName("ActionOperation")
    val sparkContext = new SparkContext(sparkConf)

    val nums : List[Int] = List(1, 2, 3, 4,5,6,7,8,9,10)

    val number = sparkContext.parallelize(nums)
    var count = number.count()
    var ints = number.take(5)
    println(count)
    ints.foreach(num => println(num))

  }
  def main(args: Array[String]): Unit = {
    /*map()*/
    count()
  }

}
